### ISSUE TITLE
Dependency cycle detection does not seem to work 

### ISSUE THREAD

**Issue by soerendomroes** [lf-lang/lingua-franca#1024]
Hi, tried to do the esweek tutorial and I noticed that the dependency cycle analysis does not work.

```
target C

reactor Printer {
    input in: int
    output out: int
    reaction(in) -> out {=
        // Using a thread-safe print function provided by the runtime.
        info_print("Hello World! %d", in->value);
    =}
}

main reactor {
    p = new Printer()
    p.out -> p.in
}
```
![tutorial](https://user-images.githubusercontent.com/6419799/157700021-994c2d03-fb83-4cbf-883e-9a8d1353c45f.svg)

No validator error is created. So it does not seem to be a problem of the diagram.

**Comment by a-sr** [comment]
I had a look at the problem because I was worried that it relates to changes done for modal models but it seems not the case.
It looks like a mishandled corner case. I was able to track down the issue to [this point in ReactionInstanceGraph](https://github.com/lf-lang/lingua-franca/blob/c344185ef296f3a0768828fc9f3e3a2b0b5e7ee9/org.lflang/src/org/lflang/generator/ReactionInstanceGraph.java#L138). In this instance `dstRuntime` and `srcRuntime` are actually the same and hence no edge is introduced.
My guess is that the code cannot handle this corner case, because as soon as there are two reactions involved, it works again:
![Screenshot from 2022-03-10 17-36-10](https://user-images.githubusercontent.com/25612037/157710742-adbf9fe5-2188-4d9d-9fd8-72a0638b4c23.png)
I don't fully understand the handling of runtime instances in this algorithm yet but the `if` was explicitly introduced [here](https://github.com/lf-lang/lingua-franca/commit/6cffccb2c35985537c93f0672f9b5a6aa1508abf) to fix a problem with federated. Maybe someone else can take it up from here.

**Comment by edwardalee** [comment]
Thanks for this pointer. I'm trying to figure out why I put in the check that lets this corner case through... There must have been a reason... :-)

**Comment by edwardalee** [comment]
I think this will be fixed by PR #1026, but let's see whether the CI tests reveal further problems. I removed the check that @a-sr identified, and this resulted in only one test breaking, `src/modal_models/ModalCycleBreaker.lf`.
This test seems to have had a genuine cycle in mode Two.
That cycle was being masked by the bug.
I removed the cycle and the test passes, but @a-sr should check that the test is still testing what was intended.
